src/app/utils/field-validation/validators/tableValidator.ts

```diff
@@ -0,0 +1,151 @@
+
+  if (addMoreRows) return right(response)
+  return answerArray.length === minimumRows
+    ? right(response)
+    : left(
+        `TableValidator:\tanswer has extra rows even though addMoreRows is false`,
+      )
+}
+
+/**
+ * Returns a validation function to check if the
+ * response has more than the maximum number of rows specified.
+ */
+const makeMaximumRowsValidator: TableValidatorConstructor = (tableField) => (
+  response,
+) => {
+  const { answerArray } = response
+  const { maximumRows } = tableField
+
+  if (!maximumRows) return right(response)
+
+  return answerArray.length <= maximumRows
+    ? right(response)
+    : left(`TableValidator:\tanswer has more than the maximum number of rows`)
+}
+
+/**
+ * Returns a validation function to check if the
+ * response has the correct number of answers for each row.
+ */
+const makeRowLengthValidator: TableValidatorConstructor = (tableField) => (
+  response,
+) => {
+  const { answerArray } = response
+  const { columns } = tableField
+
+  return answerArray.every((row) => row.length === columns.length)
+    ? right(response)
+    : left(`TableValidator:\tanswer has rows with incorrect number of answers`)
+}
+
+/**
+ * Returns a validation function that applies
+ * the correct validator for each column based on its type.
+ */
+const makeColumnValidator: TableValidatorConstructor = (tableField) => (
+  response,
+) => {
+  const { answerArray, isVisible, _id } = response
+  const { columns } = tableField
+
+  const cols = unzip(answerArray)
+  return columns.every((columnFormField, i) => {
+    return cols[i].every((answer) => {
+      if (
+        columnFormField.columnType !== BasicField.ShortText &&
+        columnFormField.columnType !== BasicField.Dropdown
+      ) {
+        return false
+      }
+
+      // Inject properties so that columnField can be passed into validateField
+      type ColumnFormFieldWithProperties<T> = T & {
+        getQuestion: { (): string }
+        description: string
+        disabled: boolean
+        fieldType: BasicField
+      }
+
+      const columnField = columnFormField
+      ;(columnField as ColumnFormFieldWithProperties<IColumnSchema>).disabled = false
+      ;(columnField as ColumnFormFieldWithProperties<IColumnSchema>).description =
+        'some description'
+      ;(columnField as ColumnFormFieldWithProperties<IColumnSchema>).fieldType =
+        columnFormField.columnType
+      ;(columnField as ColumnFormFieldWithProperties<IColumnSchema>).getQuestion = () =>
+        'some question'
+
+      const columnResponse = {
+        answer,
+        isVisible,
+        fieldType: columnFormField.columnType,
+        _id,
+        question: columnFormField.title,
+      }
+      return validateField(
+        columnField._id,
+        columnField as ColumnFormFieldWithProperties<IColumnSchema>,
+        columnResponse,
+      ).isOk()
+    })
+  })
+    ? right(response)
+    : left(`TableValidator:\tanswer failed field validation`)
+}
```

Comment:
This function is complicated and rather unsatisfactory - first the need to `unzip` to invert a row-based representation to column-based which is a fairly expensive operation, proceeded by complicated typecasting just to be able to call `validateField` at the end.

What if we continue the theme of functional programming by using composition instead? It could look something like the following:

1. Write a factory function to create an array of `IFieldSchema` object instances out of an array of `columnFormField`s, using existing mongoose schema functions such as `createShortTextField`.
2. Create a custom row validator function (instead of column validator) that uses the array of `IFieldSchema` objects to construct a validation function that can validate an entire `answerArray`
3. Table validation passes if every row in the table passes the row validator

This should make the implementation very easy to compose, and much more trivial to extend the table field to include other field types later on e.g. NRIC

Reply:
How about this implementation:

1) I defined a type `CoreField`, which is the minimum number of properties required for a field to be validated. 
```
export type CoreField = {
  required: boolean
  title: string
  _id?: string // _id is required for signature validation
  fieldType: BasicField
}
```
Then we have `DropdownCoreField`, `ShortTextCoreField` etc which are extensions of `CoreField`.

2) A column field is a `CoreField` but with additional property `columnType` and an optional `fieldType`. So before passing the column field into `validateField()`, I take `answerField.fieldType = answerField.columnType`, and then use a typeguard to check that `fieldType` property is set.

The main reason why I did it this way instead of declaring both column field and normal form fields as extension of CoreField, is that otherwise all of downstream validation functions would have to accept both column and normal form fields, and check for `columntype` and `fieldtype` respectively. Whereas by converting column fields into `CoreField` before calling validateField, the validators deal with a single consistent type.

3) Proposed implementation calls validateField on each nested table field, after checking for the column type, without the need for `unzip`. So we can extend to include other field types simply by allowing in more column types in `const ALLOWED_COLUMN_TYPES`